cpp/include/raft/random/detail/rng_impl.cuh

```diff
@@ -16,906 +16,306 @@
-                    OutType* val,
-                    ExponentialDistParams<OutType> params,
-                    LenType idx    = 0,
-                    LenType stride = 0)
+template <typename OutType, typename LenType = int>
+void exponential(
+  RngState& rng_state, OutType* ptr, LenType len, OutType lambda, cudaStream_t stream)
 {
-  OutType res;
-  gen.next(res);
-  constexpr OutType one = (OutType)1.0;
-  *val                  = -raft::myLog(one - res) / params.lambda;
+  ExponentialDistParams<OutType> params;
+  params.lambda = lambda;
+  RAFT_CALL_RNG_FUNC(rng_state, call_rng_kernel<1>, rng_state, stream, ptr, len, params);
 }
 
-template <typename GenType, typename OutType, typename LenType>
-DI void custom_next(GenType& gen,
-                    OutType* val,
-                    RayleighDistParams<OutType> params,
-                    LenType idx    = 0,
-                    LenType stride = 0)
+template <typename OutType, typename LenType = int>
+void rayleigh(RngState& rng_state, OutType* ptr, LenType len, OutType sigma, cudaStream_t stream)
 {
-  OutType res;
-  gen.next(res);
-
-  constexpr OutType one = (OutType)1.0;
-  constexpr OutType two = (OutType)2.0;
-  *val                  = raft::mySqrt(-two * raft::myLog(one - res)) * params.sigma;
+  RayleighDistParams<OutType> params;
+  params.sigma = sigma;
+  RAFT_CALL_RNG_FUNC(rng_state, call_rng_kernel<1>, rng_state, stream, ptr, len, params);
 }
 
-template <typename GenType, typename OutType, typename LenType>
-DI void custom_next(GenType& gen,
-                    OutType* val,
-                    LaplaceDistParams<OutType> params,
-                    LenType idx    = 0,
-                    LenType stride = 0)
+template <typename OutType, typename LenType = int>
+void laplace(
+  RngState& rng_state, OutType* ptr, LenType len, OutType mu, OutType scale, cudaStream_t stream)
 {
-  OutType res, out;
-
-  do {
-    gen.next(res);
-  } while (res == OutType(0.0));
-
-  constexpr OutType one     = (OutType)1.0;
-  constexpr OutType two     = (OutType)2.0;
-  constexpr OutType oneHalf = (OutType)0.5;
-
-  // The <= comparison here means, number of samples going in `if` branch are more by 1 than `else`
-  // branch. However it does not matter as for 0.5 both branches evaluate to same result.
-  if (res <= oneHalf) {
-    out = params.mu + params.scale * raft::myLog(two * res);
-  } else {
-    out = params.mu - params.scale * raft::myLog(two * (one - res));
-  }
-  *val = out;
+  LaplaceDistParams<OutType> params;
+  params.mu    = mu;
+  params.scale = scale;
+  RAFT_CALL_RNG_FUNC(rng_state, call_rng_kernel<1>, rng_state, stream, ptr, len, params);
 }
 
-template <typename GenType, typename OutType, typename LenType>
-DI void custom_next(
-  GenType& gen, OutType* val, SamplingParams<OutType, LenType> params, LenType idx, LenType stride)
+template <typename DataT, typename WeightsT, typename IdxT = int>
+void sampleWithoutReplacement(RngState& rng_state,
```

Comment:
One very small nitpick I have here is that the handle should always be the first argument for primitives that accept one.

Reply:
I removed the handle here since it wasn't being used in the implementation at all (the stream is already given as another parameter).
This also keeps the API in line with other RNG APIs